rfd/0194-k8s-rbac-simplification.md

```diff
@@ -0,0 +1,220 @@
+---
+authors: Guillaume Charmes (guillaume.charmes@goteleport.com)
+state: draft
+---
+
+# RFD 0194 - Kubernetes RBAC Simplification
+
+## Required Approvers
+
+- Engineering: @rosstimothy && @hugoShaka && @tigrato
+
+## What
+
+The purpose of this RFD is to propose a simplification of the Kubernetes
+RBAC (Role-Based Access Control) within the Teleport project.
+
+The goal is to streamline the user experience and reduce the complexity of
+getting up and running on day 1 and managing permissions later on.
+
+## Why
+
+Today, the Kubernetes RBAC in Teleport is complex and can be difficult for
+users to manage.
+A proper setup requires some external Kubernetes config and matching role
+setup.
+This complexity leads to misconfigurations and security issues.
+
+Simplifying the RBAC model will help improve usability and security.
+
+### References
+
+A common issue currently is the un-intuitive result of complex rule sets and
+more critically, difficult to get setup on day 1.
+
+- @klizhentas struggling to setup a cluster following the `self-hosted` flow:
+  (internal) <https://gravitational.slack.com/archives/C03SULRAAG3/p1715394587181739>
+- Customer confusion on role resolution:
+  - <https://github.com/gravitational/teleport/issues/45475>
+  - (internal) <https://goteleport.slack.com/archives/C0582MBSMHN/p1738187497666819>
+
+## Goals
+
+- Make the Teleport agent authoritative
+  - Handle permissions for the users instead of relying on impersonating a
+    different user/group
+- Discourage (or even deprecate?) the use of the `kubernetes_groups` and
+  `kubernetes_users` in favor of `kubernetes_resources`
+- Clarify / Improve documentation for the various ways to setup Kubernetes in
+  Teleport.
+
+## Proposal
+
+### Background
+
+Currently, when enrolling a Kubernetes cluster and setting it up to use the
+role `system:masters`, everything works as expected. Teleport is super admin
+and yields reduced permissions to the user based on the `kubernetes_resources`
+and `kubernetes_label` fields from the Teleport role.
+
+### Impersonation
+
+To limit the number of code changes, the proposal is to default the
+`kubernetes_group` value to `system:masters` and hide it from the role.
+This can be done by allowing the field to be `null`, and treating `null` as
+`systems:masters`.
+
+The `kubernetes_user` should default to the Teleport user auth email.
```

Comment:
I really like the idea of removing dimensions instead of adding yet another new feature to unify them all. This is both backward compatible, and simplifies the user experience by making them use the `kubernetes_resources` field.

Reply:
This is not backwards compatible

Reply:
This change is not backward compatible because, previously, if `kubernetes_users` and `kubernetes_groups` were unset (whether null or empty, as they are marshaled identically in proto fields), users would not have access to the cluster. However, under your proposal, users who previously had no access would now gain `system:masters` access. This introduces a security risk, breaks backward compatibility guarantees, and could lead to multiple S1 incidents.

Reply:
This would only apply to roles >= v8, it won't impact existing roles

Reply:
What if the user has role v8 and v7 assigned at the same time?

Reply:
The v7 role would behave as it does today, the v8 would have the new behavior.